File: min-stack-single-stack.py
# Time Complexity : O(1)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : No

class MinStack(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack1 = []
        self.min = float('inf')

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if x <= self.min:
            self.stack1.append(self.min)
            self.min = x
        self.stack1.append(x)
        
    def pop(self):
        """
        :rtype: None
        """
        if self.stack1[-1] == self.min:
            self.stack1.pop()
            self.min = self.stack1.pop()
        else:
            self.stack1.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stack1[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: MinStack_SingleStack.java
//Time Complexity:O(1)
//Space Complexity:O(n)
//Solution accepted and submitted on Leetcode.

import java.util.Stack;

class MinStack{
    
    private Stack<Integer> stack;
    int min;

    public MinStack() {
    	stack = new Stack<>();
    	min = Integer.MAX_VALUE;
    }
    
    public void push(int x) {
        if(x <= min)
        {
        	stack.push(min);
            min = x;
        }
        stack.push(x);    
    }
    
    public void pop() {
    	int p = stack.pop();
        if(p == min)
        {
        	min = stack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min;
    }
}
